pkg/koordlet/runtimehooks/hooks/terwayqos/terwayqos.go

```diff
@@ -0,0 +1,428 @@
+			PodName:      meta.Pod.Name,
+			PodNamespace: meta.Pod.Namespace,
+			PodUID:       string(meta.Pod.UID),
+			Prio:         getPodPrio(meta.Pod),
+			CgroupDir:    filepath.Join("/sys/fs/cgroup/net_cls", meta.CgroupDir),
+			QoSConfig: QoSConfig{
+				IngressBandwidth: ing,
+				EgressBandwidth:  egress,
+			},
+		}
+	}
+
+	p.lock.Lock()
+	p.pods = pods
+	p.lock.Unlock()
+
+	select {
+	case p.syncChan <- struct{}{}:
+	default:
+	}
+	return nil
+}
+
+func (p *Plugin) syncPodConfig() error {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	if p.enabled == nil {
+		return nil
+	}
+
+	if !*p.enabled {
+		_ = os.Remove(p.podFilePath)
+		return nil
+	}
+
+	_, err := os.Stat(p.podFilePath)
+	if err != nil {
+		_, err = os.Create(p.podFilePath)
+		if err != nil {
+			return fmt.Errorf("create terway qos pod file failed, err: %v", err)
+		}
+	}
+
+	outPods, err := json.Marshal(p.pods)
+	if err != nil {
+		return err
+	}
+
+	pEvent := audit.V(3).Node().Reason("netqos reconcile").Message("update pod to : %v", string(outPods))
+	pRes, err := resourceexecutor.NewCommonDefaultUpdater(p.podFilePath, p.podFilePath, string(outPods), pEvent)
+	if err != nil {
+		return err
+	}
+
+	p.executor.UpdateBatch(true, pRes)
+	return nil
+}
+
+func (p *Plugin) syncNodeConfig() error {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	if p.enabled == nil {
+		return nil
+	}
+
+	if !*p.enabled {
+		_ = os.Remove(p.nodeFilePath)
+		return nil
+	}
+
+	_, err := os.Stat(p.nodeFilePath)
+	if err != nil {
+		_, err = os.Create(p.nodeFilePath)
+		if err != nil {
+			return fmt.Errorf("create terway qos node file failed, err: %v", err)
+		}
+	}
+
+	outNode, err := p.node.MarshalText()
+	if err != nil {
+		return err
+	}
+
+	nEvent := audit.V(3).Node().Reason("netqos reconcile").Message("update node to : %v", string(outNode))
+	nRes, err := resourceexecutor.NewCommonDefaultUpdater(p.nodeFilePath, p.nodeFilePath, string(outNode), nEvent)
+	if err != nil {
+		return err
+	}
+
+	p.executor.UpdateBatch(true, nRes)
+	return nil
+}
+
+// parseNetQoS only LS and BE is taken into account
+func parseNetQoS(slo *slov1alpha1.NodeSLOSpec, node *Node) error {
+	if slo.SystemStrategy == nil {
+		return nil
+	}
+	if slo.SystemStrategy.TotalNetworkBandwidth.Value() < 0 {
+		return fmt.Errorf("invalid total network bandwidth %d", slo.SystemStrategy.TotalNetworkBandwidth.Value())
+	}
+	total := uint64(slo.SystemStrategy.TotalNetworkBandwidth.Value())
+
+	// to Byte/s
+	node.HwTxBpsMax = BitsToBytes(total)
+	node.HwRxBpsMax = BitsToBytes(total)
+
+	qos := slo.ResourceQOSStrategy
+	if qos == nil {
+		return nil
+	}
+
+	if qos.LSClass != nil && qos.LSClass.NetworkQOS != nil && qos.LSClass.NetworkQOS.Enable != nil && *qos.LSClass.NetworkQOS.Enable {
+		q, err := parseQoS(qos.LSClass.NetworkQOS, total)
+		if err != nil {
+			return err
+		}
+		node.L1RxBpsMin = q.IngressRequestBps
+		node.L1RxBpsMax = q.IngressLimitBps
+		node.L1TxBpsMin = q.EgressRequestBps
+		node.L1TxBpsMax = q.EgressLimitBps
+	}
+
```

Comment:
LSRClass also belongs to l1

Reply:
If all configurations are allowed, it will lead to chaos. I hope to keep only two configuration options.

Reply:
What you are referring to as "mapping" is the assignment of priorities to Pods, which is accomplished in the getPodPrio method. 
The requests and limits for L1/L2 configuration are obtained from the LSClass and BEClass definitions. It should not cause any confusion.

Reply:
it seems NetworkQOS in ResourceQOSStrategy.LSRClass should be ignored in terway qos plugin currently, since there are only limited level(l0-l2) in configurations